### PR TITLE
Book and Category services test impl

### PR DESCRIPTION
**Pull request by mkazarian-1**
Screenshot 2024-12-05 190028
Screenshot 2024-12-05 190127
Screenshot 2024-12-05 190145
Screenshot 2024-12-05 190159

### CODE REVIEW

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        sqlCommandExecutor(dataSource, "books/delete-books-with-category.sql");
+    }
+
+    @AfterEach
+    public void afterEach() {
+        sqlCommandExecutor(dataSource, "books/delete-books-with-category.sql");
+    }
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: you can use Sql annotation instead

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+        CategoryDto categoryDto3 = new CategoryDto();
+        categoryDto3.setId(3L);
+        categoryDto3.setName("lol info3");
+        categoryDto3.setDescription("some description3");
+        expected.add(categoryDto3);
+
+        CategoryDto categoryDto4 = new CategoryDto();
+        categoryDto4.setId(4L);
+        categoryDto4.setName("lol info4");
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: better move it to the private method or better util class and simplify your test cases

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        final MvcResult mvcResultGet = mockMvc.perform(
+                        get("/categories/2")
+                )
+                .andExpect(status().isOk())
+                .andReturn();
+
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: move it to the separate methods

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        MvcResult mvcResult = mockMvc.perform(
+                delete("/categories/1")
+        ).andExpect(status().isNoContent()).andReturn();
+
+        mockMvc.perform(
+                        get("/categories/1")
+                )
+                .andExpect(status().isNotFound())
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: the same. Don’t use other endpoints to test results. your test cases should be isolated

File: src/test/java/org/example/bookshop/mapper/BookMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+import org.example.bookshop.dto.book.CreateBookRequestDto;
+import org.example.bookshop.dto.book.UpdateBookRequestDto;
+import org.example.bookshop.mapper.impl.BookMapperImpl;
+import org.example.bookshop.model.Book;
+import org.example.bookshop.model.Category;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class BookMapperTest {
+    private final BookMapper bookMapper = new BookMapperImpl();
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/mapper/BookMapperTest.java]: Don’t test mapstruct ;)

File: src/test/java/org/example/bookshop/repository/BookRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    @Test
+    @DisplayName("""
+            Find all books with Pageable
+            """)
+    @Sql(scripts = "classpath:books/add-books-with-category.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = "classpath:books/delete-books-with-category.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void findAll_WithBooksAndCategories_ReturnsListOfBooks() {
+        List<Book> actual = bookRepository.findAll(
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/repository/BookRepositoryTest.java]: Usually, crud methods are not tested

File: src/test/resources/books/add-books-with-category-except-id.sql [hunk 7]
```diff
@@ -0,0 +1,7 @@
+insert into categories (id, is_deleted, name, description) values ( 1, false, 'lol info','some description');
+insert into categories (id, is_deleted, name, description) values ( 2, false, 'lol info1','some description');
+
```
Comment (liliia-ponomarenko) [inline src/test/resources/books/add-books-with-category-except-id.sql]:
```suggestion
insert into categories (id, is_deleted, name, description) values (1, false, 'lol info','some description');
insert into categories (id, is_deleted, name, description) values (2, false, 'lol info1','some description');

```

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution and test the service layer ;)

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java (hunk 1 above)
Comment (mkazarian-1) [reply src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: I think it is better write only one method then a lot of annotation for each method

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+
+    @BeforeAll
+    static void beforeAll(
+            @Autowired DataSource dataSource,
+            @Autowired WebApplicationContext applicationContext) {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        sqlCommandExecutor(dataSource, "books/delete-books-with-category.sql");
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: Answering your question you can add annotation above class with the same execution phase

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        MvcResult mvcGetResult = mockMvc
+                .perform(
+                        get("/books/1")
+                )
+                .andExpect(status().isOk())
+                .andReturn();
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]: not fixed

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+
+        System.out.println(mvcGetResult.getResponse().getContentAsString());
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]:
```suggestion
```

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+                actualGet.getAuthor());
+        Assertions.assertEquals(expected.getIsbn(),
+                actualGet.getIsbn());
+        Assertions.assertEquals(expected.getPrice()
+                        .stripTrailingZeros(),
+                actualGet.getPrice().stripTrailingZeros());
+        assertThat(actualGet.getCategoryIds())
+                .containsExactlyInAnyOrderElementsOf(
+                        expected.getCategoryIds()
+                );
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]: remove it

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+    @DisplayName("""
+            Should delete Book in DB and return code 204
+            """)
+    void deleteBook_BookExist_Success() throws Exception {
+        //When
+        mockMvc.perform(
+                        get("/books/1")
+                )
+                .andExpect(status().isOk())
+                .andReturn();
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]: the same issue with test isolation

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+                delete("/books/1")
+        ).andExpect(status().isNoContent()).andReturn();
+
+        mockMvc.perform(
+                        get("/books/1")
+                )
+                .andExpect(status().isNotFound())
+                .andReturn();
+        //Then
+        assertEquals("", mvcResult
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]: add informative message

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+
+        //Then
+        List<BookDto> actual = objectMapper.readValue(
+                mvcResult.getResponse().getContentAsString(),
+                objectMapper.getTypeFactory().constructCollectionType(List.class, BookDto.class)
+        );
+        Assertions.assertEquals(4, actual.size());
+    }
+
+    private static @NotNull List<BookDto> getBookDtoList() {
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]: do you really need this annotation?

File: src/test/java/org/example/bookshop/repository/BookRepositoryTest.java [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        expected.setTitle("Jhon Snow1");
+        expected.setAuthor("Author1");
+        expected.setIsbn("978-2388981567");
+        expected.setPrice(new BigDecimal(5));
+        expected.setDescription("""
+                sed do eiusmod tempor incididunt
+                ut labore et dolore magna aliqua.
+                """);
+
+        Book actual1 = bookRepository.save(expected);
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/repository/BookRepositoryTest.java]: Don’t  test crud methods

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes and test the service layer ;)

File: src/test/java/org/example/bookshop/controller/BookControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+
+        //Then
+        List<BookDto> actual = objectMapper.readValue(
+                mvcResult.getResponse().getContentAsString(),
+                objectMapper.getTypeFactory().constructCollectionType(List.class, BookDto.class)
+        );
+        Assertions.assertEquals(4, actual.size());
+    }
+
+    private static List<BookDto> getBookDtoList() {
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/BookControllerTest.java]: can you move it to the util class?

File: src/test/java/org/example/bookshop/controller/CategoryControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+            Should delete Category in DB and return code 204
+            """)
+    void deleteBook_BookExist_Success() throws Exception {
+        //When
+        mockMvc.perform(
+                delete("/categories/1")
+        ).andExpect(status().isNoContent()).andReturn();
+    }
+
+    private static List<CategoryDto> getDtoList() {
```
Comment (liliia-ponomarenko) [inline src/test/java/org/example/bookshop/controller/CategoryControllerTest.java]: the same

File: src/test/resources/application.properties [hunk 18]
```diff
@@ -1,8 +1,8 @@
-spring.datasource.url=jdbc:h2:mem:testdb
-spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.url=jdbc:h2:mem:///book_store
```
Comment (liliia-ponomarenko) [inline src/test/resources/application.properties:1]: Let’s  use test conteiners

File: src/test/resources/books/delete-books-with-category.sql [hunk 19]
```diff
@@ -0,0 +1,2 @@
+DELETE FROM books;
+DELETE FROM categories;
```
Comment (liliia-ponomarenko) [inline src/test/resources/books/delete-books-with-category.sql]: add one empty line. check other files.

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Let’s test the service layer!!!

**Comment by mkazarian-1** [comment]
> Let’s test the service layer!!!

I don't see the point of testing the service layer separately, since there is no special functionality there and the only thing we do is create a layer between the repository and the controller and wake up errors. Therefore, I think that integration tests are enough here

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Comment by sarakhmen** [comment]
I don't want to block you with this hw due to the huge amount of attempts. But Liliia is right here. You should test service layer using UNIT tests. It's a usual practice and there are always unit tests for services on real projects. Do it next time (you can even to it now and push changes)

**Pull request merged**